presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java

```diff
@@ -159,18 +163,34 @@ public CachingHiveMetastore(
 
     public static CachingHiveMetastore memoizeMetastore(ExtendedHiveMetastore delegate, boolean isMetastoreImpersonationEnabled, long maximumSize, int partitionCacheMaxColumnCount)
     {
-        return new CachingHiveMetastore(
-                delegate,
-                newDirectExecutorService(),
-                isMetastoreImpersonationEnabled,
-                OptionalLong.empty(),
-                OptionalLong.empty(),
-                maximumSize,
-                false,
-                ALL,
-                0.0,
-                partitionCacheMaxColumnCount,
-                NOOP_METASTORE_CACHE_STATS);
+        if (delegate instanceof AlluxioHiveMetastore) {
+            return new CachingHiveMetastore(
+                    delegate,
+                    newDirectExecutorService(),
+                    isMetastoreImpersonationEnabled,
+                    OptionalLong.empty(),
+                    OptionalLong.empty(),
+                    maximumSize,
+                    false,
+                    ALL,
+                    0.0,
+                    partitionCacheMaxColumnCount,
+                    NOOP_METASTORE_CACHE_STATS);
+        }
+        else {
+            return new CachingHiveMetastore(
+                    delegate.getDelegate().get(),
+                    newDirectExecutorService(),
+                    isMetastoreImpersonationEnabled,
+                    OptionalLong.empty(),
+                    OptionalLong.empty(),
+                    maximumSize,
+                    false,
+                    ALL,
+                    0.0,
+                    partitionCacheMaxColumnCount,
+                    NOOP_METASTORE_CACHE_STATS);
+        }
```

Comment:
Can you explain a bit what you are doing here ?

Reply:
In cases other than AlluxioHiveMetastore, the delegate getting passed to `memoizeMetastore()` method is already an instance of CachingHiveMetastore. This was resulting in CachingHiveMetastore calling CachingHiveMetastore again via its delegate, which in turn will then call the actual metastore implementation. 
Hence for cases other than AlluxioHiveMetastore, I am now doing `delegate.getDelegate().get()` so that the intended metastore is set as the delegate and there are no redundant calls to CachingHiveMetastore.
If you feel this is something not related to this PR, I can raise a new one for this and we can merge that first. I came to know about this issue when checking the number of getTable() calls. @NikhilCollooru

Reply:
The purpose of memoize metastore is to return a CachingHiveMetastore with 100% caching enabled for all calls. 
CachingHiveMetastore(scope = ALL) --calls -> normal CachingHIveMetastore(scopre=as per what is configured in the config) ---calls --> actual Metastore implementation.

The reasoning is that the redundant calls to Metastore inside one transaction should be served from cache irrespective of the actual scope configured. Meaning lets assume scope=None is configured. This means no calls will be cached inside CachingHiveMetastore and all calls will be forwarded to actual Metastore. Lets say the query is using tableA 10 times in the query (select a from tableA union select b from tableA union ..etc) .. then the memoize metastore calls will make sure that only the first getTable(tableA) goes to actual Metastore and the remaining redundant 9 getTable calls will be served from the cache. 

Your current implementation is removing the middle CachingHiveMetastore and enabling 100% caching to all Metastore calls irrespective of scope configured. It is incorrect.

Reply:
@NikhilCollooru I have pinged you on presto slack. Let us discuss this in detail there.

Reply:
@NikhilCollooru I have pinged you on slack for deeper discussions. 
Just trying to understand your example here. 

> CachingHiveMetastore(scope = ALL) (this instance is per query/transaction)--calls -> normal CachingHIveMetastore(scopre=as per what is configured in the config)(this is singleton instance that is shared by all queries) ---calls --> actual Metastore implementation.

Lets call the per query/transaction CHM as CHM1 and the singleton instance as CHM2. CHM1 gets instantiated in memoizeMetastore call.
Lets assume scope=PARTITION is configured. As per the [code](https://github.com/prestodb/presto/blob/c3dbf9a6655a36321dc595222f256f5e52889547/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java#L166), CHM1 always has ttl = 0. So in your example, tableCache object for CHM1 has ttl=0 and since scope is PARTITION, tableCache object for CHM2 also has ttl=0. So every getTable call goes to actual metastore only. 

Please let me know if I am missing anything here.

Reply:
Having said that, on a high level, I still feel 100% caching within a given transaction can be achieved with a single instance of CHM. All you need to do is only allow scope=ALL and enable CHM1 to make use of MetastoreClientConfig configs. 
Also if we ultimately want to have 100% caching enabled for all calls within a given transaction/query, then what is the point of providing the option of setting scope to PARTITION? Please let me know @NikhilCollooru

Reply:
> Lets assume scope=PARTITION is configured. As per the [code](https://github.com/prestodb/presto/blob/c3dbf9a6655a36321dc595222f256f5e52889547/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java#L166), CHM1 always has ttl = 0.
This is incorrect. OptionalLong.empty() **DOES NOT** mean ttl=0 . Please see the usage here [https://github.com/prestodb/presto/blob/master/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java#L1093](code). It means TTL is not set. In guava cache, it means that the cache entry lives until the max capacity of the cache (1000 here) is reached.
> By default cache instances created by CacheBuilder will not perform any type of eviction.
https://guava.dev/releases/17.0/api/docs/com/google/common/cache/CacheBuilder.html

Reply:
> Also if we ultimately want to have 100% caching enabled for all calls within a given transaction/query, then what is the point of providing the option of setting scope to PARTITION?

let me re-iterate cache scope is for sharing the cache across all queries. When we set the scope to PARTITION ..it means that the partition object is cache across all queries. If CHM2 does not exist ...then there is no way to share the cache partition/table object across queries .

Reply:
Thank you for the useful explanations @NikhilCollooru and thank you for being patient with my doubts. I now fully understand the use of CHM2, but I still think the overall design can be simplified. I guess providing the option of setting scope=PARTITION is redundant, because the partition caches are already taken care of with scope=ALL.

I am trying to understand the use of CHM1 now, because the possible values of scope are `ALL` and `PARTITION` only. Setting any other value for scope (for example `NONE`) results in throwing IllegalArgumentException. Consider a scenario where only CHM2 exists with scope=ALL always. Effectively if I am removing this memoizeMetastore() method completely, do you see any case where 100% caching is not enabled for all calls in a transaction?

In other words, is CHM1 needed only because we are providing the option of setting scope=PARTITION to users, or does it solve any other use case as well? This should be my final question on this discussion. Looking forward to hearing from you.

Reply:
> In other words, is CHM1 needed only because we are providing the option of setting scope=PARTITION to users, or does it solve any other use case as well?
Per-transaction metastore cache addresses the issue of repeated access of same data in the metastore. CHM1 will make sure that a given query will get consistent view of snapshot of metadata. It will make sure all the duplicate metastore calls will return the same data irrespective of whether the metadata changed on server side or not. Assume a given query calls metadataCall1 multiple times. CHM1 will make sure the result of the metadataCall1 is always the same across all the calls even if the data changed on the metastore side. 

lets say CHM1 does not exist and CHM2 has scope = NONE. Then every call goes to Metastore and its not guaranteed that the same call might return the same data..since there is no guarantee that the table/partition is not modified during the duration of the query.

Reply:
So on a high level, CHM1 is needed to cater to cases which cannot be covered with CHM2. At present, setting scope=PARTITION seems to be the only such case to me because if the scope is set to ALL for CHM2, it can only cache every call. 
To re-iterate my point, scope cannot be set to NONE as per the [code](https://github.com/prestodb/presto/blob/864619df81857c61db353af99455369452771f8f/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java#L225), and I am still not convinced on why we provide the option of setting scope as PARTITION. 
As I understand this, our goal of caching every single metadata call in a transaction is only getting complicated if some user overrides the value of scope to PARTITION ([it is by default set to ALL](https://github.com/prestodb/presto/blob/864619df81857c61db353af99455369452771f8f/presto-hive-metastore/src/main/java/com/facebook/presto/hive/MetastoreClientConfig.java#L49)). As per the documentation, guava cache does not put any constraint on the maximum size as well, so having only CHM2 in place is again not a problem. 

Hence I suggest to change this caching design of presto to only have CHM2 in place. In case of Alluxio the singleton instance is not CachingHiveMetastore, but that is all implementation details. We can maybe take it up separately if you and the community agrees. I am reverting the changes for now so we can get this PR merged. @NikhilCollooru

Reply:
Actually after thinking it through, I guess what you are saying makes sense. Both CHM1 and CHM2 are needed in place because if the scope of CHM2 is set to ALL, then all the metadata calls are cached including listTableNames, listSchemaNames in a shared cache, and this might not be the desired behaviour for many users. 

I am currently working on fixing the build and failing test cases. Once that is done, I will ping you again for another pass @NikhilCollooru